compiler/rustc_trait_selection/src/solve/project_goals.rs

```diff
@@ -116,17 +116,34 @@ impl<'tcx> assembly::GoalKind<'tcx> for ProjectionPredicate<'tcx> {
                 .map(|pred| goal.with(tcx, pred));
             ecx.add_goals(where_clause_bounds);
 
-            // In case the associated item is hidden due to specialization, we have to
-            // return ambiguity this would otherwise be incomplete, resulting in
-            // unsoundness during coherence (#105782).
-            let Some(assoc_def) = fetch_eligible_assoc_item_def(
+            let assoc_def = match fetch_eligible_assoc_item_def(
                 ecx,
                 goal.param_env,
                 goal_trait_ref,
                 goal.predicate.def_id(),
-                impl_def_id
-            )? else {
-                return ecx.evaluate_added_goals_and_make_canonical_response(Certainty::AMBIGUOUS);
+                impl_def_id,
+            ) {
+                Ok(assoc_def) => assoc_def,
+                Err(NotAvailableReason::ErrorGuaranteed(guar)) => {
+                    let error_term = ecx.term_error_of_kind(goal.predicate.term, guar);
+                    ecx.eq(goal.param_env, goal.predicate.term, error_term)
+                        .expect("expected goal term to be fully unconstrained");
+                    return ecx.evaluate_added_goals_and_make_canonical_response(Certainty::Yes);
+                }
+                // In case the associated item is hidden due to specialization, we have to
+                // return ambiguity during coherence as it would otherwise be incomplete,
+                // resulting in unsoundness (#105782).
+                //
+                // Outside of coherence we want to fail here as we want to treat defaulted
+                // associated items as opaque.
+                Err(NotAvailableReason::DefaultItem) => match ecx.solver_mode() {
+                    SolverMode::Normal => return Err(NoSolution),
+                    SolverMode::Coherence => {
+                        return ecx.evaluate_added_goals_and_make_canonical_response(
+                            Certainty::AMBIGUOUS,
+                        );
+                    }
+                },
```

Comment:
Would it even be unsound to return `Err(NoSolution)`  _inside_ the solver during coherence, its very much true that a deafult impl will never allow us to actually normalize. For normalization outside of the solver this is wrong since we normalize shit even though we dont have to so making progress based of "this is required to actually normalize instead of just be wf" would be wrong.